substratevm/mx.substratevm/testhello.py

```diff
@@ -784,13 +783,14 @@ def test():
     checker = Checker('info args', rexp)
     checker.check(exec_string)
     
-    exec_string = execute("x/i $pc")
-    if arch == 'aarch64':
-        rexp = r"%ssub%ssp, sp, #0x%s"%(wildcard_pattern, spaces_pattern, hex_digits_pattern)
-    else:
-        rexp = r"%ssub %s\$0x%s,%%rsp"%(wildcard_pattern, spaces_pattern, hex_digits_pattern)
-    checker = Checker('x/i $pc', rexp)
-    checker.check(exec_string)
+    if asm_insns:
+        exec_string = execute("x/i $pc")
+        if arch == 'aarch64':
+            rexp = r"%ssub%ssp, sp, #0x%s"%(wildcard_pattern, spaces_pattern, hex_digits_pattern)
+        else:
+            rexp = r"%ssub %s\$0x%s,%%rsp"%(wildcard_pattern, spaces_pattern, hex_digits_pattern)
+        checker = Checker('x/i $pc', rexp)
+        checker.check(exec_string)
```

Comment:
@fniephaus What was the reason for making this optional?

Reply:
This part of the test is a bit overfitting as it specifies what instructions the compiler is allowed to emit for method prologues. For example, GraalVM EE uses `-:H+StackOverflowCheckInPrologue` by default and therefore emits different instructions. Instead of making this optional, we could also remove the check altogether. After all, I believe the tests should simulate representative debugging scenarios to ensure we provide a decent user experience, without relying on specific decisions made by the compiler.

Reply:
@fniephaus I understand that we want to avoid the test depending on what the compiler does but there was a good reason why this rule was broken here. Indeed, the fact that this check caught a difference in EE means it has served its purpose ;-)

The instruction check is actually just part of a sequence of checks designed to ensure that parameter variable encodings are correct. That sequence includes a prior and subsequent validation of the output from the `info args` command. The method being debugged has been deliberately chosen so that it will have a mix of int greg and freg arguments *plus* also integer and floating arguments passed via the stack.

The break point is carefully placed at the first instruction of the method, an address guaranteed to provide var info for all params. That is where we first run `info args`. The point of the stepi instruction on x86/two stepi instructions on aarch64 is to transition to the next infopoint at the end of the method prologue, also guaranteed to provide an updated set of var info for all params. The point of testing before and after the stack push is to ensue that the stack offsets for the stack-passed parameters have been encoded correctly both before and after push

So, this means that on EE we won't have completed that stack push transition and so the second validation of the `info args` command is not going to be checking the offsets post stack-push. That does not mean that the test will fail. What it does mean is that the test has been weakened and may catch an error in the calculation of the stack offsets post-push. It would be better to come up with something that allows the pre- and post-push offset calculations to be validated on EE as well as CE.

What we need is to be able to guarantee we stop at the point where i) the stack push is complete and ii) param info is available. I can be sure there is param info available at the stack push instruction because I ensure that the generator explicitly writes info at that point. The alternative would be to stop using a break point, say at a specific line. The problem with that is that param var location info cannot be guaranteed to be available at any chosen line break -- it comes and goes according to the whim of the compiler's infopoint encoding scheme, which is far more complex than any decisions it makes regarding the stack push.

What we could perhaps do instead of stepping a fixed number of instructions is to step until we actually see a stack extend instruction, execute that instruction and only then test the params values. This would build in a slightly weaker assumption i.e. that last instruction emitted by the compiler in a method prologue is a stack adjustment. That's not too presumptious (after all we already assume a fixed size frame). Making this work means we would need to detect any instruction on the current CPU that adjusts the stack register. Does that sound like a good idea? @olpaw any thoughts on this?

Reply:
Thanks for the shedding some light on this, @adinn. Your suggestion seems very reasonable to me, so I have removed the option for making the check optional and replaced the logic that skips over the method prologue with a new one that steps instructions until the first that adjusts the stack pointer register (see https://github.com/oracle/graal/commit/66d1486241a1aa84aab79624a4937f79f42e4cb4). I hope I've documented this well enough in the test code.

Reply:
That looks good but I think it still needs a small adjustment. If you look at the previous test code you will see that the aarch64 branch executes two `stepi` commands after he check for a stack push. That's because on aarch64 the stack build sequence requires both a stack extend and then a save of sp to fp. The compiler only posts a PROLOGUE_END notification after the second instruction. This also means that the param var locations in the debug info are only redefined after the move instruction. So, if you don't do the 2nd stepi then the pre-push var locations will be used. So, I would expect this version of the test to fail on aarch64.

Reply:
Doh! I think this should fixes it: https://github.com/oracle/graal/pull/5691/commits/fc40db371521d883068d4bf819f9765a88c1dbfe#diff-911ff5ce4770f7bdd9410ad3374ae2d3f4fc7fbc69b661cac4b2d7077062ee1cR805-R813.

Reply:
Yes, looks good.